nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java

```diff
@@ -303,20 +310,24 @@ private RenderingContext getRealContext(RenderingContext ctx) {
      * @since 7.2
      */
     private Object getSingletonInstance(RenderingContext ctx) {
-        if (singleton == null) {
-            singleton = getNewInstance(ctx, true);
-        } else {
-            for (Field contextField : contextFields) {
-                ThreadSafeRenderingContext value;
-                try {
-                    value = (ThreadSafeRenderingContext) contextField.get(singleton);
-                } catch (IllegalArgumentException | IllegalAccessException e) {
-                    log.error("unable to create a marshaller instance for clazz " + clazz.getName(), e);
-                    return null;
+        if (!servicesInjected) {
+            synchronized (this) {
+                if (!servicesInjected) {
+                    injectServices(singleton);
+                    servicesInjected = true;
                 }
-                value.configureThread(ctx);
             }
         }
+        for (Field contextField : contextFields) {
+            ThreadSafeRenderingContext value;
+            try {
+                value = (ThreadSafeRenderingContext) contextField.get(singleton);
+            } catch (IllegalArgumentException | IllegalAccessException e) {
+                log.error("unable to create a marshaller instance for clazz " + clazz.getName(), e);
```

Comment:
Here and elsewhere I'd rather rethrow than return null, unless there's a good (documented) reason for invalid classes to be used.